include/deal.II/base/tensor.h

```diff
@@ -1313,7 +1313,7 @@ template <typename ArrayLike, std::size_t... indices>
 constexpr DEAL_II_HOST_DEVICE_ALWAYS_INLINE
 Tensor<rank_, dim, Number>::Tensor(const ArrayLike &initializer,
                                    std::index_sequence<indices...>)
-  : values{{Tensor<rank_ - 1, dim, Number>(initializer[indices])...}}
+  : values{{value_type(initializer[indices])...}}
```

Comment:
Since this shows up in #16511, let me explain this change.

It used to be that `values` was an array (C-style, but that is not important here) of objects of type `Tensor<rank-1,dim,Number>`, and so the cast of the initializer elements to that type made sense.

But since #16480, we use an array of `Number` objects for the case `rank==1`, whereas it continues to be an array of `Tensor<rank-1,dim,Number` for `rank>1`:
```
std::conditional_t<rank_ == 1,
                     std::array<Number, dim>,
                     std::array<Tensor<rank_ - 1, dim, Number>, dim>>
    values;
```
That means tha casting the initializers to `Tensor<rank-1,dim,Number>` is wrong if `rank==1`. Rather, the right thing to do is to cast it to the type used in the array, which is `value_type`:
```
  using value_type =
    std::conditional_t<rank_ == 1, Number, Tensor<rank_ - 1, dim, Number>>;
```
The fact that after #16480 and before this patch, the existing code continued to work is that we can initialize the `values` array with a list of initializers that can be converted to the type of the array, and we have `Tensor<0,dim,Number>::operator Number &();`. In other words, for `rank==1`, the *old* code first created a `Tensor<0,dim,Number>` via the initializers, and then converted it back to `Number` right away.

This seemed silly, and this patch was intended to avoid the circuitous logic.

Reply:
The failure in #16511 looks like this:
```
/srv/temp/testsuite-mA3DEm67/dealii/include/deal.II/base/tensor.h:1316:13: error: cannot convert 'const value_type' (aka 'const Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>>') to 'value_type' (aka 'double') without a conversion operator
  : values{{value_type(initializer[indices])...}}
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/srv/temp/testsuite-mA3DEm67/dealii/include/deal.II/base/tensor.h:1467:5: note: in instantiation of function template specialization 'dealii::Tensor<1, 2>::Tensor<dealii::Tensor<1, 2, Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>>>, 0UL, 1UL>' requested here
  : Tensor(initializer, std::make_index_sequence<dim>{})
    ^
/srv/temp/testsuite-mA3DEm67/dealii/include/deal.II/base/tensor.h:1316:13: note: in instantiation of function template specialization 'dealii::Tensor<1, 2>::Tensor<Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>>>' requested here
  : values{{value_type(initializer[indices])...}}
            ^
```
The error message is not particularly helpful because it apparently uses `value_type` for two different types. But what I *think* is happening is that we are trying to instantiate the current function, which now look like
```
template <int rank_, int dim, typename Number>
template <typename ArrayLike, std::size_t... indices>
constexpr DEAL_II_HOST_DEVICE_ALWAYS_INLINE
Tensor<rank_, dim, Number>::Tensor(const ArrayLike &initializer,
                                   std::index_sequence<indices...>)
  : values{{value_type(initializer[indices])...}}
{ [...]
```
with `Number=doube` and `initializer` of type `Tensor<1, 2, Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>>>`. So `initializer[indices]` should return a `const Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>>` object, which apparently cannot be converted to `double`. That's what the error message says:
```
/srv/temp/testsuite-mA3DEm67/dealii/include/deal.II/base/tensor.h:1316:13: error: cannot convert 'const value_type' (aka 'const Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>>') to 'value_type' (aka 'double') without a conversion operator
```

What is unclear to me is how this *used* to work before this patch. With the *old* code, we first initialized a `Tensor<0,dim,double>` from the `Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>>'`, which was then implicitly converted to `double`. The first step would have happened with this constructor
```
template <int dim, typename Number>
template <typename OtherNumber>
constexpr DEAL_II_HOST_DEVICE_ALWAYS_INLINE
Tensor<0, dim, Number>::Tensor(const OtherNumber &initializer)
  : value(internal::NumberType<Number>::value(initializer))
{}
```
and the second via
```
template <int dim, typename Number>
constexpr DEAL_II_HOST_DEVICE_ALWAYS_INLINE
Tensor<0, dim, Number>::operator Number &()
{
  Assert(dim != 0,
         ExcMessage("Cannot access an object of type Tensor<0,0,Number>"));
  return value;
}
```
This would suggest that the following sequence of calls should work for the constructor we're currently looking at:
```
template <int rank_, int dim, typename Number>
template <typename ArrayLike, std::size_t... indices>
constexpr DEAL_II_HOST_DEVICE_ALWAYS_INLINE
Tensor<rank_, dim, Number>::Tensor(const ArrayLike &initializer,
                                   std::index_sequence<indices...>)
  : values{{(internal::NumberType<Number>::value(initializer[indices]))...}}
```
But this only triggers the same error again:
```
1: /home/bangerth/p/deal.II/1/dealii/include/deal.II/base/tensor.h: In instantiation of ‘constexpr dealii::Tensor<rank_, dim, Number>::Tensor(const
 ArrayLike&, std::index_sequence<_Ind ...>) [with ArrayLike = dealii::Tensor<2, 2, Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<do
uble, double> > >; long unsigned int ...Indices = {0, 1}; int rank_ = 2; int dim = 2; Number = double; std::index_sequence<_Ind ...> = std::integer
_sequence<long unsigned int, 0, 1>]’:
1: /home/bangerth/p/deal.II/1/dealii/include/deal.II/base/tensor.h:1467:56:   required from ‘constexpr dealii::Tensor<rank_, dim, Number>::Tensor(c
onst dealii::Tensor<rank_, dim, OtherNumber>&) [with OtherNumber = Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double, double> >;
 int rank_ = 2; int dim = 2; Number = double]’
1: /home/bangerth/p/deal.II/1/dealii/include/deal.II/physics/elasticity/kinematics.h:391:24:   required from ‘dealii::Tensor<2, dim, Number> dealii
::Physics::Elasticity::Kinematics::w(const dealii::Tensor<2, dim, Number>&, const dealii::Tensor<2, dim, Number>&) [with int dim = 2; Number = Saca
do::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double, double> >]’
1: /home/bangerth/p/deal.II/1/dealii/tests/sacado/../ad_common_tests/physics_functions_01.h:74:39:   required from ‘void test_physics() [with int d
im = 2; number_t = double; dealii::Differentiation::AD::NumberTypes ad_type_code = dealii::Differentiation::AD::NumberTypes::sacado_dfad]’
1: /home/bangerth/p/deal.II/1/dealii/tests/sacado/physics_functions_01_1.cc:34:58:   required from here
1: /home/bangerth/p/deal.II/1/dealii/include/deal.II/base/tensor.h:1316:49: error: no matching function for call to ‘dealii::internal::NumberType<d
ouble>::value(const value_type&)’
1:  1316 |   : values{{(internal::NumberType<Number>::value(initializer[indices]))...}}
1:       |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
```
I have to admit that I am confused, because if this sequence of things is not possible I don't understand how it worked before...